spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/TraceReactorAutoConfiguration.java

```diff
@@ -194,27 +213,122 @@ public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory)
+			addQueueWrapper(springContext);
+		}
+		else {
+			boolean decorateOnEach = environment.getProperty(
+					"spring.sleuth.reactor.decorate-on-each", Boolean.class, true);
+			if (decorateOnEach) {
+				if (log.isTraceEnabled()) {
+					log.trace("Decorating onEach operator instrumentation");
+				}
+				Hooks.onEachOperator(SLEUTH_TRACE_REACTOR_KEY,
+						scopePassingSpanOperator(springContext));
 			}
-			Hooks.onLastOperator(SLEUTH_TRACE_REACTOR_KEY,
-					scopePassingSpanOperator(springContext));
+			else {
+				if (log.isTraceEnabled()) {
+					log.trace("Decorating onLast operator instrumentation");
+				}
+				Hooks.onLastOperator(SLEUTH_TRACE_REACTOR_KEY,
+						scopePassingSpanOperator(springContext));
+			}
+		}
+		decorateScheduler(springContext);
+	}
+
+	static void addQueueWrapper(ConfigurableApplicationContext springContext) {
+		Hooks.addQueueWrapper(SLEUTH_TRACE_REACTOR_KEY,
+				queue -> traceQueue(springContext, queue));
+	}
+
+	@Override
+	public void close() throws IOException {
+		if (log.isTraceEnabled()) {
+			log.trace("Cleaning up hooks");
 		}
-		Schedulers.setExecutorServiceDecorator(
+		Hooks.resetOnEachOperator(SLEUTH_TRACE_REACTOR_KEY);
+		Hooks.resetOnLastOperator(SLEUTH_TRACE_REACTOR_KEY);
+		Schedulers.resetOnScheduleHook(
+				TraceReactorAutoConfiguration.SLEUTH_REACTOR_EXECUTOR_SERVICE_KEY);
+	}
+
+	private static void decorateScheduler(ConfigurableApplicationContext springContext) {
+		Schedulers.onScheduleHook(
 				TraceReactorAutoConfiguration.SLEUTH_REACTOR_EXECUTOR_SERVICE_KEY,
-				(scheduler,
-						scheduledExecutorService) -> new TraceableScheduledExecutorService(
-								springContext, scheduledExecutorService));
+				ReactorSleuth.scopePassingOnScheduleHook(springContext));
+	}
+
+	private static Queue<?> traceQueue(ConfigurableApplicationContext springContext,
+			Queue<?> queue) {
+		if (!springContext.isActive()) {
+			return queue;
+		}
+		CurrentTraceContext currentTraceContext = springContext
+				.getBean(CurrentTraceContext.class);
+		@SuppressWarnings("unchecked")
+		Queue<Envelope> envelopeQueue = (Queue<Envelope>) queue;
+		return new AbstractQueue<Object>() {
+
+			@Override
+			public int size() {
+				return envelopeQueue.size();
+			}
+
+			@Override
+			public boolean offer(Object o) {
+				TraceContext traceContext = currentTraceContext.get();
+				return envelopeQueue.offer(new Envelope(o, traceContext));
+			}
+
+			@Override
+			public Object poll() {
+				Envelope envelope = envelopeQueue.poll();
+				if (envelope == null) {
+					return null;
+				}
+				if (envelope.traceContext != null) {
+					// puts in thread local
+					currentTraceContext.maybeScope(envelope.traceContext);
+				}
+				return envelope.body;
+			}
+
+			@Override
+			public Object peek() {
+				return queue.peek();
+			}
+
+			@Override
+			@SuppressWarnings("unchecked")
+			public Iterator<Object> iterator() {
+				return (Iterator<Object>) queue.iterator();
+			}
```

Comment:
be careful with these two. They are not currently used by Reactor (I _think_), but there is no guarantee that they won't be used in future, and it may happen that the object offered to this queue won't be the same as `peek`ed, so I would highly recommend *at least* adding the unwrapping to these

Reply:
+1